File: packages/plugin-sdk/assembly/schedule.ts
import {
  peekoo_schedule_cancel,
  peekoo_schedule_get,
  peekoo_schedule_set,
} from "./host";
import { extractBoolField, extractRawField, extractStringField, extractU64Field, quote } from "./json";
import { readString, writeString } from "./memory";
import { ScheduleInfo } from "./types";

/**
 * Create or replace a schedule timer.
 */
export function set(
  key: string,
  interval_secs: u64,
  repeat: bool,
  delay_secs: u64 = 0,
): void {
  let input = "{\"key\":" + quote(key) + ",\"interval_secs\":" + interval_secs.toString() + ",\"repeat\":" + (repeat ? "true" : "false");
  if (delay_secs > 0) {
    input += ",\"delay_secs\":" + delay_secs.toString();
  }
  input += "}";
  peekoo_schedule_set(writeString(input));
}

/**
 * Cancel a schedule timer.
 */
export function cancel(key: string): void {
  const input = "{\"key\":" + quote(key) + "}";
  peekoo_schedule_cancel(writeString(input));
}

/**
 * Get information about a schedule timer.
 * Returns null if no timer with this key exists.
 */
export function get(key: string): ScheduleInfo | null {
  const input = "{\"key\":" + quote(key) + "}";
  const offset = peekoo_schedule_get(writeString(input));
  const response = readString(offset);

  const scheduleJson = extractRawField(response, "schedule");
  if (scheduleJson.length == 0 || scheduleJson == "null") {
    return null;
  }

  const info = new ScheduleInfo();
  info.owner = extractStringField(scheduleJson, "owner");
  info.key = extractStringField(scheduleJson, "key");
  info.interval_secs = extractU64Field(scheduleJson, "interval_secs");
  info.repeat = extractBoolField(scheduleJson, "repeat");
  info.time_remaining_secs = extractU64Field(scheduleJson, "time_remaining_secs");

  return info;
}

File: crates/peekoo-plugin-sdk/src/schedule.rs
//! Timer scheduling.
//!
//! Requires the `scheduler` permission. Scheduled timers fire
//! `schedule:fired` events with `{ "key": "<your_key>" }` in the payload.
//!
//! ```no_run
//! use peekoo_plugin_sdk::prelude::*;
//!
//! fn example() -> Result<(), Error> {
//!     peekoo::schedule::set("my_timer", 300, true, None)?;
//!     if let Some(info) = peekoo::schedule::get("my_timer")? {
//!         peekoo::log::info(&format!("{}s remaining", info.time_remaining_secs));
//!     }
//!     peekoo::schedule::cancel("my_timer")?;
//!     Ok(())
//! }
//! ```

use extism_pdk::{Error, Json};

use crate::host_fns::{
    peekoo_schedule_cancel, peekoo_schedule_get, peekoo_schedule_set, ScheduleCancelRequest,
    ScheduleGetRequest, ScheduleSetRequest,
};
use crate::types::ScheduleInfo;

/// Create or replace a schedule timer.
///
/// - `key` — unique identifier for this timer (per plugin).
/// - `interval_secs` — interval in seconds between firings.
/// - `repeat` — if `true`, the timer repeats; if `false`, it fires once.
/// - `delay_secs` — optional initial delay before the first firing.
///   Defaults to `interval_secs` when `None`.
pub fn set(
    key: &str,
    interval_secs: u64,
    repeat: bool,
    delay_secs: Option<u64>,
) -> Result<(), Error> {
    unsafe {
        peekoo_schedule_set(Json(ScheduleSetRequest {
            key: key.to_string(),
            interval_secs,
            repeat,
            delay_secs,
        }))?;
    }
    Ok(())
}

/// Cancel a schedule timer.
pub fn cancel(key: &str) -> Result<(), Error> {
    unsafe {
        peekoo_schedule_cancel(Json(ScheduleCancelRequest {
            key: key.to_string(),
        }))?;
    }
    Ok(())
}

/// Get information about a schedule timer.
///
/// Returns `None` if no timer with this key exists.
pub fn get(key: &str) -> Result<Option<ScheduleInfo>, Error> {
    let response = unsafe {
        peekoo_schedule_get(Json(ScheduleGetRequest {
            key: key.to_string(),
        }))?
    };

    Ok(response.0.schedule)
}
